compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -1007,6 +1007,83 @@ class NamedOpConversion : public OpRewritePattern<OpTy> {
   SmallVector<int64_t> permutation;
 };
 
+// Fuses a transpose into a reduction linalg.generic by absorbing it into the
+// indexing map.
+class FuseTransposeThroughGenericReduction
+    : public OpRewritePattern<linalg::GenericOp> {
+public:
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(genericOp)) {
+      return failure();
+    }
+
+    // Only try to fuse when the generic has at least one reduction dimension.
+    if (genericOp.getNumParallelLoops() == genericOp.getNumLoops()) {
+      return rewriter.notifyMatchFailure(genericOp, "not a reduction");
+    }
+
+    // All maps must be projected permutations.
+    if (!llvm::all_of(genericOp.getIndexingMapsArray(), [](AffineMap map) {
+          return map.isProjectedPermutation();
+        })) {
+      return rewriter.notifyMatchFailure(genericOp,
+                                         "not a projected permutation");
+    }
+
+    // Look for a transpose on any input.
+    for (int64_t inputIdx = 0; inputIdx < genericOp.getNumDpsInputs();
```

Comment:
nit: don't re-evaluate the end value

```suggestion
    for (int64_t inputIdx = 0, endIdx = genericOp.getNumDpsInputs(); inputIdx < endIdx
```